### ISSUE TITLE
Assign expression parsing

### ISSUE THREAD

**Issue by mayakwd** [HaxeFoundation/intellij-haxe#624]
During rework of semantic core I've faced with weird interpretion of ASSIGN_EXPRESSION and math expressions, like ADDITIVE_EXPRESSION

Some examples:
1. Math operation have higher priority than assign expression:
Psi structure of ```i = i + 4;``` interpreting as 
```
ADDITIVE_EXPRESSION (i = i + 4)
   ASSIGN EXPRESSION (i = i)
      REFERENE_EXPRESSION (i)
      ...
      REFERENCE_EXPRESSION (i)
   ...
   LITERAL_EXPRESSION (4)
```
It should be: 

```
ASSIGN_EXPRESSION
   REFERENCE_EXPRESSION (i)
   ...
   ADDITIVE_EXPRESSION (i+4)
```

For such construction - everything is fine:
Psi structure of ```i = (i + 4);``` interpreting as 
```
ASSIGN_EXPRESSION (i = (i + 4))
   REFERENE_EXPRESSION (i)
   ...
   PARENTHESIZED_EXPRESSION
      ADDITIVE EXPRESSION (i + 4)
```

2. Psi structure of ```i = i2 = 4;``` interpreting as
```
ASSIGN_EXPRESSION (i = i2 = 4)
   ASSIGN_EXPRESSION (i = i2)
   ...
   LITERAL_EXPRESSION 4
```
It should be:
```
ASSIGN_EXPRESSION (i = i2 = 4)
   REFERENCE_EXPRESSION (i)
   ...
   ASSIGN_EXPRESSION (i2 = 4)
      REFERENCE_EXPRESSION (i2)
      ...
      LITERAL_EXPRESSION (4)
```

**Comment by EricBishton** [comment]
I did this on purpose.  The JFlex-based lexer doesn't understand generic (type parameter) syntax as different from regular operators, and it cannot communicate with the parser, and cannot make the determination whether it should be returning operators or termination sequences for generic types.  This causes issues with returning the correct tokens.  For example, `var arr: Array<Map<String,String>> = ...`  In this case, the `>>=` is always incorrectly detected by the lexer as SHIFT_RIGHT_AND_ASSIGN.  The solution was to disallow the lexer gathering the largest possible token, and let the parser get the individual characters as tokens and do the matching.  Correct ordering/priority of the operators was the next problem, because the parser would always see the greater-than operator before any of the shift operators.  

Old issues were #42 and #287, PR was #576.  The solution at the time was to change the ordering for parsing.  That made it so that we could at least parse the files without error.  Now that you actually want to do a mock evaluation, that requires a correct solution.

**Comment by mayakwd** [comment]
Well in this case I guess we should adopt hxparser for our purposes, and rewrite a lot of code. It's not that easy, but I think much hackish code will be removed.

**Comment by EricBishton** [comment]
I would prefer to leave this open until/unless we decide not to tackle the issue at all.  I think this is a valid issue that needs addressing.

**Comment by mayakwd** [comment]
Fixed with #637
